ethylene-mapper/src/main/java/com/github/steanky/ethylene/mapper/type/Token.java

```diff
@@ -0,0 +1,654 @@
+     * @param typeVariables the parameters to use
+     * @return a new token containing the result of parameterizing this type
+     * @throws IllegalArgumentException if the given TypeVariableMap does not contain some necessary types
+     */
+    public final @NotNull Token<?> parameterizeWithOwner(@NotNull Token<?> owner,
+        @NotNull TypeVariableMap typeVariables) {
+        Objects.requireNonNull(owner);
+        Objects.requireNonNull(typeVariables);
+
+        Class<?> raw = rawType();
+        TypeVariable<?>[] parameters = raw.getTypeParameters();
+        return parameterizeInternal(raw, owner.get(), extractTypeArgumentsFrom(typeVariables.resolve(), parameters));
+    }
+
+    /**
+     * Convenience overload for {@link Token#parameterizeWithOwner(Token, TypeVariableMap)}.
+     *
+     * @param owner         the owner class
+     * @param typeVariables the parameters to use
+     * @return a new token containing the result of parameterizing this type
+     * @throws IllegalArgumentException if the given TypeVariableMap does not contain some necessary types
+     */
+    public final @NotNull Token<?> parameterizeWithOwner(@NotNull Class<?> owner,
+        @NotNull TypeVariableMap typeVariables) {
+        Objects.requireNonNull(owner);
+        Objects.requireNonNull(typeVariables);
+
+        Class<?> raw = rawType();
+        TypeVariable<?>[] parameters = raw.getTypeParameters();
+        return parameterizeInternal(raw, owner, extractTypeArgumentsFrom(typeVariables.resolve(), parameters));
+    }
+
+    /**
+     * Creates a new token whose type is an array, whose component type is the current type of this token. If the type
+     * represented by this token is a raw class, the returned token will contain the result of calling
+     * {@link Class#arrayType()}. Otherwise, an appropriate implementation of {@link GenericArrayType} will be used
+     * instead.
+     *
+     * @return a new token representing an array type
+     */
+    public final @NotNull Token<?> arrayType() {
+        Type type = get();
+        if (type instanceof Class<?> cls) {
+            return Token.ofType(cls.arrayType());
+        }
+
+        return Token.ofType(GenericInfo.resolveType(new WeakGenericArrayType(type)));
+    }
+
+    /**
+     * Creates a new token whose type is the component type of the array type represented by this token. If this token
+     * does not contain an array type, an {@link IllegalStateException} is thrown.
+     *
+     * @return a token containing the component type of the array type underlying this token
+     */
+    public final @NotNull Token<?> componentType() {
+        Type type = get();
+        if (!TypeUtils.isArrayType(type)) {
+            throw new IllegalStateException("This token does not represent an array type");
+        }
+
+        Type component = TypeUtils.getArrayComponentType(type);
+        ClassLoader loader = ReflectionUtils.getClassLoader(component);
```

Comment:
unused variable? if the method call serves a purpose, I would create a separate method to assert that a class loader exists